Enable matrix-only solve type for Eigenvalue executioner

fdkong opened the pull request:
Closes #18493

fdkong committed:
Implemented a new solver option PJFNKMO

where we use MatMult to replace expensive residual evaluations

Closes #18493

fdkong committed:
Added test for solver option PJFNKMO

Closes #18493

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/790499/) on 4f9ac72 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/18527/site)

This comment will be updated on new commits.

YaqiWang commented:
@fdkong I will get to this soon. Thanks for this speedy work.

YaqiWang commented on framework/src/utils/SlepcSupport.C:
```diff
@@ -36,7 +40,7 @@ InputParameters
 {
   MooseEnum solve_type("POWER ARNOLDI KRYLOVSCHUR JACOBI_DAVIDSON "
-                       "NONLINEAR_POWER NEWTON PJFNK JFNK",
+                       "NONLINEAR_POWER NEWTON PJFNK PJFNKMO JFNK",
```
I do not like this name because this is not Jacobian free in sense that we assemble both A and B matrix. Do we have a test for `NEWTON`? If we do not we should add one. I personally think `NEWTON` is not that useful and suggest to name this type as `NEWTON`. Or we can name this type as `MONK` (monk? ;-)) or `PMONK`.

fdkong replied:
I like the name I have right now

Not sure if we should keep arguing on this

YaqiWang commented on framework/src/utils/SlepcSupport.C:
```diff
@@ -48,7 +52,8 @@ getSlepcValidParams(InputParameters & params)
                       "PJFNK: Preconditioned Jacobian-free Newton-Kyrlov"
-                      "JFNK: Jacobian-free Newton-Kyrlov");
+                      "JFNK: Jacobian-free Newton-Kyrlov"
+                      "PJFNKMO: Preconditioned Jacobian-free Newton-Kyrlov with matrix only");
```
Preconditioned matrix only Newton-Krylov.

YaqiWang commented on framework/src/utils/SlepcSupport.C:
```diff
@@ -217,6 +222,13 @@ setEigenProblemSolverParams(EigenProblem & eigen_problem, const InputParameters
     eigen_problem.solverParams()._eigen_matrix_free = true;
     eigen_problem.solverParams()._precond_matrix_free = true;
   }
+  // We indeed matrices so that we can implement residual evaluations
```
indeed -> need.

YaqiWang commented on test/tests/problems/eigen_problem/jfnk_mo/tests:
```diff
@@ -0,0 +1,14 @@
+    slepc = true
+    slepc_version = '>=3.11.0'
+    expect_out = 'PJFNK\s+with\s+matrix\s+only'
+    requirement = "The system shall support use of matrix-vector multiplication as residual evaluation"
```
add `for eigenvalue calculations`.

YaqiWang reviewed (changes requested):
I will need to test this with Griffin.

YaqiWang commented on framework/src/utils/SlepcSupport.C:
```diff
@@ -532,6 +551,35 @@ moosePetscSNESFormMatricesTags(
   eigen_problem->computeMatricesTags(*sys.current_local_solution.get(), jacobians, tags);
 }
 
+PetscErrorCode
```
@fdkong running Griffin shows that we are missing
```
if (!eigen_problem->constJacobian())```
```
above Line 548.

YaqiWang commented:
Another thing I noticed is that `PJFNKMO` is using the default preconditioner with `ilu` even I set up a `Preconditioning` block in the input. I guess we do not have `JFNKMO`?

YaqiWang commented:
@fdkong looks like we did not turn on full jacobian evaluation for this solve type. Here is a backtrace:

```
    frame #1: 0x000000010e15e95e libmoose-dbg.0.dylib`ComputeJacobianThread::computeJacobian(this=0x00007ffeefbfb0e0) at ComputeJacobianThread.C:66:17
    frame #2: 0x000000010e161b55 libmoose-dbg.0.dylib`ComputeJacobianThread::onElement(this=0x00007ffeefbfb0e0, elem=0x000000011ce5adc0) at ComputeJacobianThread.C:237:3
    frame #3: 0x000000010d865f48 libmoose-dbg.0.dylib`ThreadedElementLoopBase<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> >::operator(this=0x00007ffeefbfb0e0, range=0x000000011ebbafc0, bypass_threading=false)(libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> const&, bool) at ThreadedElementLoopBase.h:217:9
    frame #4: 0x000000010e259acc libmoose-dbg.0.dylib`void libMesh::Threads::parallel_reduce<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*>, ComputeJacobianThread>(range=0x000000011ebbafc0, body=0x00007ffeefbfb0e0) at threads_pthread.h:380:5
    frame #5: 0x000000010e2577b2 libmoose-dbg.0.dylib`NonlinearSystemBase::computeJacobianInternal(this=0x000000011e008c18, tags=size=2) at NonlinearSystemBase.C:2499:9
    frame #6: 0x000000010e25bf07 libmoose-dbg.0.dylib`NonlinearSystemBase::computeJacobianTags(this=0x000000011e008c18, tags=size=2) at NonlinearSystemBase.C:2763:5
    frame #7: 0x000000010d9f42ae libmoose-dbg.0.dylib`FEProblemBase::computeJacobianTags(this=0x000000011e006418, tags=size=2) at FEProblemBase.C:5660:10
    frame #8: 0x000000010d9a64b8 libmoose-dbg.0.dylib`EigenProblem::computeMatricesTags(this=0x000000011e006418, soln=0x000000011ce9db30, jacobians=size=2, tags=size=2) at EigenProblem.C:208:3
    frame #9: 0x000000010e8341f5 libmoose-dbg.0.dylib`Moose::SlepcSupport::moosePetscSNESFormMatricesTags((null)=0x000000011c040a20, x=0x000000011c04dc20, mats=size=2, ctx=0x000000011e006418, tags=size=2) at SlepcSupport.C:552:18
    frame #10: 0x000000010e834cf2 libmoose-dbg.0.dylib`Moose::SlepcSupport::mooseSlepcEigenFormJacobianA(snes=0x000000011c040a20, x=0x000000011c04dc20, jac=0x000000011c056220, pc=0x000000011c02cc20, ctx=0x000000011e006418) at SlepcSupport.C:625:5
    frame #11: 0x0000000119711b86 libpetsc.3.15.dylib`SNESComputeJacobian(snes=0x000000011c040a20, X=0x000000011c04dc20, A=0x000000011c056220, B=0x000000011c02cc20) at snes.c:2713:10 [opt]
    frame #12: 0x00000001196d56ea libpetsc.3.15.dylib`SNESSolve_NEWTONLS(snes=0x000000011c040a20) at ls.c:222:12 [opt]
    frame #13: 0x00000001197170c4 libpetsc.3.15.dylib`SNESSolve(snes=0x000000011c040a20, b=0x000000011c04fe20, x=<unavailable>) at snes.c:4653:12 [opt]
    frame #14: 0x00000001188d04a8 libslepc.3.15.dylib`EPSSolve_Power [inlined] EPSPowerApply_SNES(eps=0x000000011c03aa20, x=<unavailable>, y=<unavailable>) at power.c:317:12 [opt]
    frame #15: 0x00000001188d048d libslepc.3.15.dylib`EPSSolve_Power(eps=0x000000011c03aa20) at power.c:429 [opt]
    frame #16: 0x00000001188c5a2b libslepc.3.15.dylib`EPSSolve(eps=0x000000011c03aa20) at epssolve.c:149:10 [opt]
    frame #17: 0x00000001152bfdf8 libmesh_dbg.0.dylib`libMesh::SlepcEigenSolver<double>::_solve_generalized_helper(this=0x000000011ce9e000, mat_A=0x000000011c02cc20, mat_B=0x000000011c032820, precond=0x0000000000000000, nev=1, ncv=3, tol=0.01, m_its=10000) at slepc_eigen_solver.C:687:10
    frame #18: 0x00000001152bf8da libmesh_dbg.0.dylib`libMesh::SlepcEigenSolver<double>::solve_generalized(this=0x000000011ce9e000, matrix_A_in=0x000000011ebc4d90, matrix_B_in=0x000000011ebc4e90, nev=1, ncv=3, tol=0.01, m_its=10000) at slepc_eigen_solver.C:397:10
    frame #19: 0x00000001152f221b libmesh_dbg.0.dylib`libMesh::EigenSystem::solve(this=0x000000011ce9d500) at eigen_system.C:267:34
    frame #20: 0x000000010e233cc6 libmoose-dbg.0.dylib`NonlinearEigenSystem::solve(this=0x000000011e008c18) at NonlinearEigenSystem.C:235:14
    frame #21: 0x000000010d9a98c7 libmoose-dbg.0.dylib`EigenProblem::doFreeNonlinearPowerIterations(this=0x000000011e006418, free_power_iterations=4) at EigenProblem.C:491:8
    frame #22: 0x000000010d9a9c17 libmoose-dbg.0.dylib`EigenProblem::solve(this=0x000000011e006418) at EigenProblem.C:536:9
    frame #23: 0x000000010d423522 libmoose-dbg.0.dylib`FEProblemSolve::solve(this=0x000000011ea9d4a8) at FEProblemSolve.C:226:14
```
You can see we did not call `ComputeFullJacobianThread` but `ComputeJacobianThread` here. I guess because of this, I cannot make this solve type work for one test in Griffin.

YaqiWang commented:
Another thing I forgot mentioning: when we compute left and right residuals even we pass in zero x, we may have nonzero residuals. We need to save these two residuals if they are nonzero and add them later as Ax - Bx/|Bx| + x_left - x_right/|Bx|. Please consult @loganharbour or @zachmprince if having questions.

fdkong committed:
Added a test for a coupled problem

fdkong committed:
Handle constant Jacobian matrices

fdkong committed:
Added test for non-homogeneous problems

YaqiWang reviewed (dismissed):
This looks good to me now. The non-homogeneous source on eigen side is not my priority. Custom preconditioning is something we are missing with `NEWTON` for solving nonlinear problem anyway. Using the full `A` matrix for preconditioning should be fine most of times.

fdkong committed:
Reset matrix state for EPS and ST

fdkong committed:
Fixed a few typos

fdkong committed:
Losen nl relative tol for ne_coupled_mo.i because we over solved it

fdkong commented:
It is ready for review @lindsayad 

@YaqiWang already agreed with the changes

lindsayad commented on framework/include/utils/SlepcSupport.h:
```diff
@@ -209,6 +209,11 @@ PetscErrorCode PCApply_MoosePC(PC pc, Vec x, Vec y);
+/**
+ * Function call for MFFD
+ */
+PetscErrorCode mooseSlepcEigenFormFunctionMFFD(void * ctx, Vec x, Vec r);
```
Beautiful name. I think we should get some more characters in here 😄

lindsayad commented on framework/src/systems/NonlinearEigenSystem.C:
```diff
@@ -383,6 +378,18 @@ NonlinearEigenSystem::getSNES()
+      cast_ptr<SlepcEigenSolver<Number> *>(&(*_eigen_sys.eigen_solver));
+
+  if (!solver)
+    mooseError("There is no a eigen solver");
```
```suggestion
    mooseError("Unable to retrieve eigen solver");
```

lindsayad commented on framework/src/utils/Conversion.C:
```diff
@@ -446,6 +447,33 @@ stringify(const SolveType & t)
+{
+  switch (t)
+  {
+    case EST_POWER:
+      return "Power";
+    case EST_ARNOLDI:
+      return "ARNOLDI";
+    case EST_KRYLOVSCHUR:
+      return "KRYLOVSCHUR";
+    case EST_JACOBI_DAVIDSON:
+      return "Jacobi Davidson";
+    case EST_NONLINEAR_POWER:
+      return "Nonlinear Power";
+    case EST_PJFNKMO:
+      return "PJFNK with matrix only";
+    case EST_NEWTON:
+      return "Newton";
+    case EST_JFNK:
+      return "JFNK";
+    case EST_PJFNK:
+      return "Preconditioned JFNK";
```
Seems like we should be consistent about casing here. Some are all caps while others are not

fdkong replied:
I see. Will change that

lindsayad commented on framework/src/utils/SlepcSupport.C:
```diff
@@ -19,11 +19,15 @@
 #include "FEProblemBase.h"
 #include "NonlinearEigenSystem.h"
-
+// libMesh
```
```suggestion
```

lindsayad commented on framework/src/utils/SlepcSupport.C:
```diff
@@ -19,11 +19,15 @@
 #include "libmesh/petsc_matrix.h"
 #include "libmesh/slepc_macro.h"
 #include "libmesh/auto_ptr.h"
+// PETSc
+#include "petscsnes.h"
+// SLEPc
```
I do not think we need the comments myself. I think the include names speak for themselves. But I do not care too much about this

fdkong replied:
Hmmm

fdkong replied:
Hmmm

lindsayad commented on framework/src/utils/SlepcSupport.C:
```diff
@@ -217,6 +222,16 @@ setEigenProblemSolverParams(EigenProblem & eigen_problem, const InputParameters
+    eigen_problem.solverParams()._precond_matrix_free = false;
+    eigen_problem.solverParams()._eigen_matrix_vector_mult = true;
+    // By default, we need to form full matrices, otherwise residual
+    // evaluations will be not accurate
```
```suggestion
    // evaluations will not be accurate
```

lindsayad commented on framework/src/utils/SlepcSupport.C:
```diff
@@ -526,12 +548,42 @@ moosePetscSNESFormMatricesTags(
+  ierr = SNESGetFunction(snes, NULL, &func, &fctx);
+  CHKERRQ(ierr);
+  if (fctx != ctx)
+  {
+    SETERRQ(
+        PetscObjectComm((PetscObject)snes), PETSC_ERR_ARG_INCOMP, "Contexts are not consistent \n");
+  }
```
Is there some RAII stuff happening behind the scenes here? I never know what might be happening with macros. If not, then we can remove braces?

fdkong replied:
I will check

lindsayad commented on framework/src/utils/SlepcSupport.C:
```diff
@@ -550,6 +602,47 @@ mooseSlepcEigenFormJacobianA(SNES snes, Vec x, Mat jac, Mat pc, void * ctx)
+    ST st;
+    Mat A, B;
+    PetscBool aisshell, bisshell;
+    // Rest ST state so that we can restrieve matrices
```
```suggestion
    // Rest ST state so that we can retrieve matrices
```

lindsayad commented on framework/src/utils/SlepcSupport.C:
```diff
@@ -550,6 +602,47 @@ mooseSlepcEigenFormJacobianA(SNES snes, Vec x, Mat jac, Mat pc, void * ctx)
+    ierr = PetscObjectTypeCompare((PetscObject)B, MATSHELL, &bisshell);
+    CHKERRQ(ierr);
+    if (aisshell || bisshell)
+    {
+      SETERRQ(PetscObjectComm((PetscObject)snes),
+              PETSC_ERR_ARG_INCOMP,
+              "A and B matrices can not be shell matrices when using PJFNKMO \n");
+    }
```
same question as above about RAII and extra braces

lindsayad commented on framework/src/utils/SlepcSupport.C:
```diff
@@ -696,12 +810,34 @@ mooseSlepcEigenFormFunctionA(SNES snes, Vec x, Vec r, void * ctx)
+    CHKERRQ(ierr);
+  }
+  else
+  {
+    moosePetscSNESFormFunction(snes, x, r, ctx, eigen_nl.eigenVectorTag());
+  }
```
```suggestion
    moosePetscSNESFormFunction(snes, x, r, ctx, eigen_nl.eigenVectorTag());
```

lindsayad commented on test/tests/problems/eigen_problem/jfnk_mo/tests:
```diff
@@ -0,0 +1,44 @@
+    csvdiff = 'ne_coupled_eigenvalues_0001.csv'
+    slepc = true
+    slepc_version = '>=3.11.0'
+    requirement = "The system shall support use of matrix-vector multiplication as residual evaluation for coupled eigenvalue problems"
```
What would happen if a user did not have `full = true` in their preconditioning block? Would we error? Because then the residual evaluation that this is trying to replace would not be accurate, correct?

fdkong replied:
It depends. Users might have a slow convergence (or even divergence), but answer will be wrong because we have Jacobian/residual on every nonlinear iteration

lindsayad reviewed (changes requested):
I saw a lot of beautiful PETSc/SLEPc code here. Mostly minor questions

fdkong commented on framework/include/utils/MooseTypes.h:
```diff
@@ -680,8 +680,9 @@ enum EigenSolveType
-  EST_PJFNK, ///< Preconditioned Jacobian-free Newton Krylov
-  EST_JFNK   ///< Jacobian-free Newton Krylov
+  EST_PJFNK,   ///< Preconditioned Jacobian-free Newton Krylov
+  EST_PJFNKMO, ///< The same as PJFNK except that matrix-vector multiplication is employed to replace residual evaluation in linear solver
```
MO: Matrix-Only

lindsayad replied:
So Jacobian-Free Matrix-Only ... I have to say that seems like a natural contradiction (although I understand that you are using matrix P and no matrix J). Is Jacobian-Free Matrix-Only terminology that is present in literature?

lindsayad replied:
It is good that we've used the Jacobian-free terminology from the start, as opposed to matrix-free, otherwise this new solve type would be Matrix-Free Matrix-Only 😆

fdkong replied:
> Is Jacobian-Free Matrix-Only terminology that is present in literature?
 
 No. We just come up this crazy name even though I do not really like it but I do not have a better one 
 
 Maybe I should call  it FastPJFNK

lindsayad replied:
Haha no, I think the current name conveys more accurate meaning than `FastPJFNK` (even if you have to think about it)

YaqiWang commented:
@fdkong looks like @lindsayad 's comments are not non-trivial. Can you address them so that we can get this merged? I'd like to add a few tests in Griffin for this as well.

jortensi commented:
We need this capability very soon. When will it be available?

fdkong commented on framework/include/problems/EigenProblem.h:
```diff
@@ -189,6 +194,16 @@ class EigenProblem : public FEProblemBase
    */
   void outputInverseEigenvalue(bool inverse) { _output_inverse_eigenvalue = inverse; }
 
+  /**
+   * Whether or not we are in a linear solver iteration
+   */
+  bool onLinearSolver() const { return _on_linear_solver; }
+
+  /**
+   * Set a flag to indicate whether or not we are in a linear solver iteration
+   */
+  void onLinearSolver(bool ols) { _on_linear_solver = ols; }
```
I would be tricky to do that since we use SNES from different path, but I will explore it a bit more on this

fdkong committed:
Update framework/src/systems/NonlinearEigenSystem.C

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

Update framework/src/utils/SlepcSupport.C

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

Update framework/src/utils/SlepcSupport.C

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

Update framework/src/utils/SlepcSupport.C

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

Update framework/src/utils/SlepcSupport.C

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

fdkong commented:
I just realized I did not make doc for the eigen value system. I should do in this PR

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/894691/) on 1ba3bd7 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>ba19d2</th>
      <th colspan=4>#18527 1ba3bd</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>81.95%</td>
      <td>82.00%</td>
      <td>+0.04%</td>
      <td>87.50%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>72274</td>
      <td>72412</td>
      <td>+138</td>
      <td>119</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15915</td>
      <td>15899</td>
      <td>-16</td>
      <td>17</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/18527/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/18527/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/18527/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/18527/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/18527/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/18527/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/18527/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/18527/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/18527/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/18527/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/18527/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/18527/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/18527/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/18527/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/18527/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/18527/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/18527/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/18527/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/18527/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/18527/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/18527/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/18527/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/18527/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/18527/coverage/tensor_mechanics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/18527/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 87.50% is less than the suggested 90.0%



This comment will be updated on new commits.

fdkong committed:
Support constant matrices for PFJNKMO

fdkong committed:
Added tests for constant matrices

fdkong commented:
It is ready to test on griffin side @YaqiWang 

You might not waste time on reviewing the code yet since I still need to polish a bit. I need to add more docs.

The basic idea is right. If "`consant_matrices = true`", then we do not involve any residual evaluations from solver. 

Only one residual evaluation will happen in  `EigenProblem::preScaleEigenVector`

```
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
    frame #0: 0x0000000102e1affb libmoose-dbg.0.dylib`EigenProblem::computeResidualTag(this=0x000000011401dc18, soln=0x0000000115333e80, residual=0x00000001153368c0, tag=3) at EigenProblem.C:268:3
   265 	                                 NumericVector<Number> & residual,
   266 	                                 TagID tag)
   267 	{
-> 268 	  TIME_SECTION("computeResidualTag", 3);
   269 	
   270 	  // Disassociate the default tags because we will associate vectors with only the
   271 	  // specific system tags that we need for this instance
Target 0: (moose_test-dbg) stopped.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
  * frame #0: 0x0000000102e1affb libmoose-dbg.0.dylib`EigenProblem::computeResidualTag(this=0x000000011401dc18, soln=0x0000000115333e80, residual=0x00000001153368c0, tag=3) at EigenProblem.C:268:3
    frame #1: 0x0000000102da9a5c libmoose-dbg.0.dylib`EigenProblem::preScaleEigenVector(this=0x000000011401dc18) at EigenProblem.C:402:3
    frame #2: 0x0000000102e1d061 libmoose-dbg.0.dylib`EigenProblem::execute(this=0x000000011401dc18, exec_type=0x00000001042ae870) at EigenProblem.C:143:5
    frame #3: 0x0000000102e1881d libmoose-dbg.0.dylib`FEProblemBase::initialSetup(this=0x000000011401dc18) at FEProblemBase.C:1032:5
    frame #4: 0x00000001027250f4 libmoose-dbg.0.dylib`Eigenvalue::init(this=0x0000000115355368) at Eigenvalue.C:177:18
    frame #5: 0x0000000103eb9f06 libmoose-dbg.0.dylib`MooseApp::executeExecutioner(this=0x0000000114017e18) at MooseApp.C:1074:19
    frame #6: 0x00000001007bc28e libmoose_test-dbg.0.dylib`MooseTestApp::executeExecutioner(this=0x0000000114017e18) at MooseTestApp.C:77:13
    frame #7: 0x0000000103eb3c48 libmoose-dbg.0.dylib`MooseApp::run(this=0x0000000114017e18) at MooseApp.C:1460:5
    frame #8: 0x0000000100008fb5 moose_test-dbg`main(argc=4, argv=0x00007ffeefbfdf20) at main.C:33:8
    frame #9: 0x00007fff7277fcc9 libdyld.dylib`start + 1
```

YaqiWang commented:
Thanks @fdkong I will pull this to give a try.

YaqiWang commented:
This is much better. I will push a test created by @jortensi  in a Griffin MR and tag you there. There are still two residual evaluations, but I am already much happier.

fdkong commented:
>There are still two residual evaluations

Where did they come from ? I would not expect any residual evaluation except the one from `EigenProblem::preScaleEigenVector`

YaqiWang commented:
> > There are still two residual evaluations
> 
> Where did they come from ? I would not expect any residual evaluation except the one from `EigenProblem::preScaleEigenVector`

They might be all from there, one through the EigenProblem::execute on initial, another is from EigenProblem::solve.

YaqiWang commented:
Also the multiple element loops in one residual evaluation hurt in that test case. But we can deal with that later.

fdkong commented:
> Also the multiple element loops in one residual evaluation hurt in that test case

Is it something unique in Eigenvalue Executioner? Or is it general for nonlinear solver as well? 

I can not believe I do that in my Eigenvalue Executioner

fdkong commented:
>They might be all from there, one through the EigenProblem::execute on initial, another is from EigenProblem::solve.

I think you suggested that for Picard iteration or something like that (I can not remember)

Do we still need that if the matrices are constant? I could easily turn that off if we do not that

YaqiWang commented:
> > Also the multiple element loops in one residual evaluation hurt in that test case
> 
> Is it something unique in Eigenvalue Executioner? Or is it general for nonlinear solver as well?
> 

It is general. When we evaluate residual, we first evaluate all pre_aux uos, then aux, then post_aux uos, then kernels. That tests contains both aux and uo evaluations. I am not sure if there are both pre_aux and post_aux in the test. Need to look into it to see.

YaqiWang commented:
> > They might be all from there, one through the EigenProblem::execute on initial, another is from EigenProblem::solve.
> 
> I think you suggested that for Picard iteration or something like that (I can not remember)
> 
> Do we still need that if the matrices are constant? I could easily turn that off if we do not need that

Yes, I did. Not sure how to avoid that and still make the EigenProblem::solve function re-entering safe.

YaqiWang commented:
If we do not use ||Bx||_2 for normalization, but let us say just ||x||_2, we can avoid residual evaluation.

YaqiWang commented:
I confirmed that that test has both pre_aux and post_aux. There is a way to change that test to only has post_aux, and we can clearly see the impact by doing that. @fdkong I made some comments in that MR. We can have conversations there.

YaqiWang commented:
Last comment for today: We need to make `solve_type=LINEAR` for non-eigenvalue problems to support this (maybe it already did).

fdkong committed:
Reset matrices flag for picard iteration

So that we rebuild matrices for each picard iteration

fdkong commented:
@GiudGiud 

Something like this should work

```
[Executioner]
  type = Eigenvalue
  solve_type = PJFNKMO
  constant_matrices = true
[]
```

Please test it out, and let me know how is going. Do not review the Code at this point. I am still missing docs

YaqiWang commented:
A reminder: we want this to be merged for Griffin soon. Please make this with a high priority. Thanks.

lindsayad commented:
@fdkong it looks like you wanted to add more docs before we reviewed? Or should we review now?

fdkong commented:
I am still working on docs. I will ping you once it is ready.

thanks for asking

YaqiWang commented:
@fdkong I hope this is your top priority for now. If not, please let me know.

fdkong committed:
Added a short doc for eigen problem

We may add more in the future

fdkong committed:
Fixed a few typos

moosebuild commented:
Job [Application Documentation](https://civet.inl.gov/job/928892/) on 7203504 : invalidated by @fdkong

Dump conda

fdkong commented:
@lindsayad @GiudGiud It is ready for review

fdkong committed:
Add Guillaume's words back

lindsayad commented on framework/doc/content/source/problems/EigenProblem.md:
```diff
@@ -2,16 +2,109 @@
+
+- `PJFNKMO`- Matrix-only Preconditioned Jacobian-free Newton Krylov: Mathematically,
+ this option is the same as PJFNK. The difference is the residual evaluation.
+ PJFNK does `real` residual evaluations, including all finite element calculations.
```
`real` is confusing, especially in code formatting. It makes me think of `Real` residual evaluations, and I know that we are not yet supporting complex number evaluation so I assume that it is `Real` 😄 So it might be better to say something like "PJFNK calls the user residual evaluation routine" (if I'm understanding things correctly)
```suggestion
 PJFNK calls the user residual/function evaluation routine, which includes all finite element calculations.
```

lindsayad commented on framework/doc/content/source/problems/EigenProblem.md:
```diff
@@ -2,16 +2,109 @@
+ this option is the same as PJFNK. The difference is the residual evaluation.
+ PJFNK does `real` residual evaluations, including all finite element calculations.
+ PJFNKMO forms residual vectors via  matrix-vector multiplications: $A (x) = Ax$
+ and $B(x) = Bx$. This is useful when matrixes are constant. This option will be
```
```suggestion
 and $B(x) = Bx$. This is useful when matrices are constant. This option will be
```

lindsayad reviewed (dismissed):
doc comments. Otherwise I think I am ready to go with this if @YaqiWang is

fdkong committed:
Update framework/doc/content/source/problems/EigenProblem.md

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

Update framework/doc/content/source/problems/EigenProblem.md

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

YaqiWang commented:
Yeah, let us get this merged. I am fine with the documentation.

GiudGiud commented on framework/doc/content/source/problems/EigenProblem.md:
```diff
@@ -2,16 +2,109 @@
+It offers linear and nonlinear eigensolvers by leveraging [SLEPc](https://slepc.upv.es)
+capabilities (mainly the EPS system). For linear problems,
+the system supports one eigenvalue and multiple eigenvalues. For
+nonlinear problems, we currently support solving for the minimum
```
```suggestion
nonlinear problems, we currently support solving for the smallest (in magnitude)
```

GiudGiud replied:
I may be wrong on this

fdkong replied:
We can call either way. I think your change is fine as well.

GiudGiud commented on framework/doc/content/source/problems/EigenProblem.md:
```diff
@@ -2,16 +2,109 @@
+
+Note that to use the linear solvers from SLEPc, we need to form accurate
+$A$ and $B$ that will be utilized to represent the underlying physics.
+If $A% and $B$ are incorrect, you will likely end with the
```
```suggestion
If $A$ and $B$ are incorrect, you will likely end up with the
```

GiudGiud commented on framework/doc/content/source/problems/EigenProblem.md:
```diff
@@ -2,16 +2,109 @@
+Details of all the linear eigensolvers can be found in
+[SLEPc Users Manual](https://slepc.upv.es/documentation/slepc.pdf).
+
+In a multiphysics environment, physics often is nonlinear. For instance,
```
```suggestion
In a multiphysics environment, physics are often nonlinear. For instance,
```

GiudGiud commented on framework/doc/content/source/problems/EigenProblem.md:
```diff
@@ -2,16 +2,109 @@
+[SLEPc Users Manual](https://slepc.upv.es/documentation/slepc.pdf).
+
+In a multiphysics environment, physics often is nonlinear. For instance,
+ neutron transport with temperature feedback. In this case,
+  $A$ and $B$ are not constant anymore; instead, they depend on
+   the eigenvector directly or indirectly. To describe our nonlinear
+    solver, we write a generalized nonlinear eigenvalue problem as,
```
```suggestion
 neutron transport with temperature feedback. In this case,
 $A$ and $B$ are not constant anymore; instead, they depend on
 the eigenvector directly or indirectly. To describe our nonlinear
 solver, we write a generalized nonlinear eigenvalue problem as,
```

GiudGiud commented on framework/doc/content/source/problems/EigenProblem.md:
```diff
@@ -2,16 +2,109 @@
+Newton can be applied to solve this problem. The only issue is that
+the derivatives of $\lambda$ with respective to $x$ can be dense.
+We use the Jacobian-free Newton method to overcome this difficulty,
+where the Jacobian is computed via finite difference.
```
```suggestion
where the action of the Jacobian is approximated via finite differences.
```

GiudGiud commented on framework/doc/content/source/problems/EigenProblem.md:
```diff
@@ -2,16 +2,109 @@
+in this system. There is a current list:
+
+- `PJFNK` - Preconditioned Jacobian-free Newton Krylov: The preconditioning matrix
+  is formed using nonegein kernels. We do have an option to allow users to
```
```suggestion
  is formed using noneigen kernels. We do have an option to allow users to
```

GiudGiud commented on framework/doc/content/source/problems/EigenProblem.md:
```diff
@@ -2,16 +2,109 @@
+          block=Executioner Preconditioning
+          link=False
+
+- `PJFNKMO`- Matrix-only Preconditioned Jacobian-free Newton Krylov: Mathematically,
```
so this is 'linear-PJFNK' essentially?

GiudGiud replied:
like the eigen-problem is linear?

fdkong replied:
Yes, correctly.

fdkong replied:
But it can do more than that.  If you have a nonlinear problem, this option allow you run more efficient residual evaluation via matrix-vector operations in  the linear step. But it could impact the convergence. 

There is a trade-off between iteration and compute time. It will be problem dependent

GiudGiud commented on framework/doc/content/source/problems/EigenProblem.md:
```diff
@@ -2,16 +2,109 @@
+           link=False
+
+- `Newton` - Newton method: Both Jacobian and preconditioning matrices are $A$.
+ This option is added for the consistency between the nonlinear system and
```
```suggestion
 This option is added for consistency between the nonlinear system and
```

GiudGiud reviewed (commented):
these are greast docs for the eigenproblem!

fdkong committed:
Incorporate Guillaume's comments

Update framework/doc/content/source/problems/EigenProblem.md

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

Update framework/doc/content/source/problems/EigenProblem.md

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

Update framework/doc/content/source/problems/EigenProblem.md

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

Update framework/doc/content/source/problems/EigenProblem.md

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

Update framework/doc/content/source/problems/EigenProblem.md

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

Update framework/doc/content/source/problems/EigenProblem.md

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

Update framework/doc/content/source/problems/EigenProblem.md

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

loganharbour commented:
I've used this pretty extensively now - big fan

YaqiWang commented on framework/doc/content/source/problems/EigenProblem.md:
```diff
@@ -2,16 +2,109 @@
+\begin{equation}
+A(x) = \lambda B(x),
+\end{equation}
+where $A(\cdot)$ and $B$ are nonlinear functions. In the EigenProblem
```
$B$ => $B(\cdot)$.

YaqiWang commented on framework/doc/content/source/problems/EigenProblem.md:
```diff
@@ -2,16 +2,109 @@
+where $A(\cdot)$ and $B$ are nonlinear functions. In the EigenProblem
+system, the kernels of $A(\cdot)$ are referred to as "noneigen,"
+and the kernels of $B(\cdot)$ are referred to as "eigen."
+To utilize Newton to solve this problem, we define $\lambda = \frac{1}{|B(x)|}$,
```
Newton => Newton method (to differentiate Newton solve type).

YaqiWang replied:
we define => we let.

Also add: this condition serves as a normalization condition as well to ensure the uniqueness of the eigenvector.

YaqiWang commented on framework/doc/content/source/problems/EigenProblem.md:
```diff
@@ -2,16 +2,109 @@
+Newton can be applied to solve this problem. The only issue is that
+the derivatives of $\lambda$ with respective to $x$ can be dense.
+We use the Jacobian-free Newton method to overcome this difficulty,
+where the action of the Jacobian is approximated via finite differences.
```
action of the Jacobian => the production of Jacobian and a solution vector

YaqiWang commented on framework/doc/content/source/problems/EigenProblem.md:
```diff
@@ -2,16 +2,109 @@
+
+- `JFNK` - Jacobian-free Newton Krylov:  Users are responsible for
+  providing a preconditioner. This allows users to build a customized
+  preconditioner, such as a sweeper,
```
such as a sweeper => such as using a sweeper in radiation transport

YaqiWang commented on framework/doc/content/source/problems/EigenProblem.md:
```diff
@@ -2,16 +2,109 @@
+ PJFNK calls the user residual/function evaluation routine, which includes all finite element calculations.
+ PJFNKMO forms residual vectors via  matrix-vector multiplications: $A (x) = Ax$
+ and $B(x) = Bx$. This is useful when matrices are constant. This option will be
+ more efficient. If the matrices are not constant, please do not use this option;
```
Replace this sentence with: If the problem is nonlinear, i.e. the matrices are not constant, you should not use this solve option because we do not attempt to re-evaluate $A$ and $B$ at each Newton iteration. This solve option is valid with Picard iterations when the matrices are constant at each Picard iteration with conditions only changing with Picard iterations. We do re-evaluate the two matrices at the beginning of each Picard iteration.

YaqiWang commented on framework/doc/content/source/problems/EigenProblem.md:
```diff
@@ -2,16 +2,109 @@
+ This option is added for consistency between the nonlinear system and
+ the nonlinear eigensystem. Since this option does not account for the derivative
+ of eigenvalue with respective to the eigenvector, this option is not efficient.
+ Users should not use it in general.
```
I think in this case, Newton iteration is essentially the inverse power iteration. If so, you may want to add this.

fdkong replied:
This is a good point. The observation  inspired me think more  on the inverse power iteration.  I am not sure this case is  the inverse power iteration. I think it is very different from the inverse power iteration.   I need to do math to figure it out.

But at this point, let us do not add that conclusion .

YaqiWang replied:
ok to me. If we are doing this `x = xold + A^-1 * residual`, we are doing inverse power iteration.

YaqiWang reviewed (changes requested):
Since @loganharbour added me as a reviewer, I provided some more comments in the docs.

fdkong committed:
Add Yaqi's comment

fdkong commented:
@loganharbour @GiudGiud @lindsayad 

I would like to get this in ASAP if you do not have more comments. So can anyone approve it?

jortensi commented:
Is this finally merged?

YaqiWang commented:
> Is this finally merged?

Merged in MOOSE but have not made its way into Griffin yet. It will take a little more time to get into bluecrab.
